report/41_Lessons-Evolution-&-Refactoring.md

```diff
@@ -1 +1,97 @@
+The team settled on using a template built with Model-View-Controller architecture in mind, along with using ASP.NET Razor template pages for providing the HTML templates.
+
+Next part was to convert the Flask template pages to a format that Razor understands, which was simple enough, though tedious.
+Learning the syntax of Razor and Flask at the same time provided somewhat of a challenge, and porting to a MVC architecture at the same time didn't make this transition easier.
+
+The arguably biggest part was the implementation of Entity Framework (EF) and porting to PostgreSQL.
+For each part of the model, an entity would have to be established in code, and a context and repository layer for accessing and changing that entity would have to be written.
+Then, on top of that, an API layer was necessary to access and mutate this state.
+On the bright side, this completely eliminated plain text SQL queries in the code itself.
+
+If such a large refactor was to be done some other time, then a more incremental approach might be better.
+If the database was first changed to rely on another DBMS, and _then_ changed to work with EF, then the refactor could become more incremental and thus less monolithic.
+The only problem here is that it would equate to doing double the work, and when a refactor has to be done in a small timeframe, then double work doesn't seem practical.
+
+#### Make sure it all works
+
+In order to ensure that the refactor didn't break anything, tests could be useful.
+Only problem is that the previous tests were an API test script along with Python unit tests.
+
+In order to see immediate results, a Github Action was created to automate running the API test.
+That didn't really work, as it didn't catch all edge cases, so instead the project got configured to be tested with the API simpulator.
+
+Implementation of a proper test suite came later down the line, with
+The team was always aware of the fact that there should be more tests, and were even reminded when things broke in production (like when the register page broke completely).
+Though it is a lesson that the team has already learned in previous projects, here the team learned it once again.
+
+### Evolution
+
+Once the necessary refactoring steps were done, the team was capable of expanding the original implementation.
+
+#### Automation is king
+
+Early on in the development process, not many automatic steps were set up in the project pipeline.
+This made evolution much harder than it had to be, as tests would have to be manually run all the time to verify builds, and releases would have to be done manually.
+
+In the pipeline, the team has focused on having many actions done automatically, as mentioned in [TODO: 2.3: CI/CD](23_Process-CI-CD-Description.md).
+This was, at first, because the course demanded it, but once the team started using automatic steps, then the value was apparent.
```

Comment:
Using the team all over the place seems quite unnatural and just a copy paste replacement for "we" or "our". Rather you reword the sentence to have it not be there.

Reply:
Trimmed from 23 to 3, if you can eliminate the rest the go ahead

Reply:
The last are fine.